### ISSUE TITLE
pytest_bdd.exceptions.GivenAlreadyUsed with step arguments

### ISSUE THREAD

**Issue by The-Compiler** [pytest-dev/pytest-bdd#157]
From the readme:

> Often it's possible to reuse steps giving them a parameter(s). This allows to have single implementation and multiple use, so less code. Also opens the possibility to use same step twice in single scenario and with different arguments!

But this doesn't seem to be the case for `@given`. My usecase is this feature file:

``` cucumber
Feature: Going back and forward.
    Testing the :back/:forward commands.

    Scenario: Going back
        Given I open backforward/1.html
        And I open backforward/2.html
        When I run :back
        Then backforward/1.html should be loaded
```

and this python file:

``` python
import pytest_bdd as bdd

bdd.scenarios('.')


@bdd.given(bdd.parsers.parse("I set {sect} -> {opt} to {value}"))
def set_setting(quteproc, sect, opt, value):
    quteproc.set_setting(sect, opt, value)


@bdd.given(bdd.parsers.parse("I open {path}"))
def open_path(quteproc, path):
    quteproc.open_path(path)


@bdd.when(bdd.parsers.parse("I run {command}"))
def run_command(quteproc, command):
    quteproc.send_cmd(command)


@bdd.then(bdd.parsers.parse("{} should be loaded"))
def url_should_be_loaded(httpbin, path):
    requests = httpbin.get_requests()
    assert requests[-1] == [('GET', url)]
```

So opening those two pages is clearly "setup", i.e. I think they both belong to `Given ...`.

But I get:

```
pytest_bdd.exceptions.GivenAlreadyUsed: Fixture "open_path" that implements this "I open backforward/2.html" given step has been already used.
```

**Comment by olegpidsadnyi** [comment]
The idea behind "given i have an A" is that there's some fixture that implements A and stores it's instance for later steps. Following steps could use it as "When I go to the A's edit page" which means you would get fixture as a dependency.
This way - there's only one instance of A. Basically saying "given I have an A" multiple times doesn't make sense. In my opinion you can't take incremental side effect action in the given section.
In your case you say "I open" and this is an "exercise" by definition and exercises belong to the when section.

**Comment by The-Compiler** [comment]
I see, that makes sense - thanks for your explanation!

**Comment by spinus** [comment]
If I occasionally need to use same given multiple times  I use container. Example:
```
@pytest.fixture
def basket():
  return {}

@given("There is {var}")
def there_is_var(basket, var):
  basket[var] = "something"
  return basket

# than I can use

@then("make sure A and B are in basket")
def make_sure(basket):
  assert "A" in basket
  assert "B" in basket
```

**Comment by olegpidsadnyi** [comment]
For the number of objects we use a convention of the "first" and "second" if there's common parent and "other" if not.
The only thing you need for another key in the pytest request is the name of the pythonic symbol.
So if you have a given step or a fixture myobject and you want to define another "implementation" of that you can simply do:

second_myobject = myobject

then you'll be able to address it as a fixture.

**Comment by The-Compiler** [comment]
There just was someone in the `#pylib` IRC channel who wanted to do the same (using `Given` for arbitrary setup steps, even if they are based on side-effects. They used [freshen](https://github.com/rlisagor/freshen) for nose before, and there that apparently worked without issues.

To be honest, I'd still also like to do the same - using Given/When/Then for setup/testing/teardown.

@olegpidsadnyi - would you be okay with a patch to turn that check off? I'm reopening this for now :smiling:

**Comment by xordoquy** [comment]
Let me jump in - it was me on IRC.
My use case was that I wanted to double check a couple of precondition before running the full scenario.
I load database dump within the Background section and want to ensure a few values to make sure that if someone altered the DB dump it doesn't affect the scenario.

**Comment by olegpidsadnyi** [comment]
@xordoquy if we are talking about some implicit setup like database which is not really related to the test, but assuming the things work maybe you shouldn't make it a part of the feature file?
It is all about just enough specification. All the preconditions could be implicit fixtures/dependencies of your test suite, if you are not specifically observe then in your then section you don't have to mention it in your specification.
The same way If you say Given my book has 500 pages in theory you shouldn't mention: Given I have a book, And I have an author of a book, And the author has pen and paper....

**Comment by olegpidsadnyi** [comment]
For the preconditions I encourage using the underlying pytest powers and inject them as a dependency. I don't believe it is a part of the feature specification, but again - rather scripting which gherkin is not designed for.
I also wonder what kind of the database import do you have. It means you can only affect it by applying some side effects in the given steps as modifications to existing data. Which is also counter dependency injection pattern and also slower.
I'm using factories in order to create objects that I only need per test scenario and only when I need them. Maybe there's a better way you do your setup?

**Comment by xordoquy** [comment]
The database is used be another application which we interact with.
There are a lot of base data that are mostly out of our control and we came to the conclusion that factories weren't the best choice as we can't guarantee the generated data will match what the application does.

Back on the topic, I need to think a bit more about what is part of the scenario and what part should be hidden in pytest. At the moment, I'm porting the test suite from nose/freshen to pytest/pytest-bdd and this is the major issue I'm facing so far.

**Comment by xordoquy** [comment]
Note that taken from the background documentation:

    Background:
        Given a global administrator named "Greg"
        And a blog named "Greg's anti-tax rants"
        And a customer named "Wilson"
        And a blog named "Expensive Therapy" owned by "Wilson"

I don't see any reason this could be turned into:

    Background:
        Given a global administrator named "Greg"
        And a blog named "Greg's anti-tax rants" owned by "Greg"
        And a customer named "Wilson"
        And a blog named "Expensive Therapy" owned by "Wilson"

or that you have a step that matches:

    r'a (?P<role>[\w ]+) named "(?P<username>\w+)"'

**Comment by olegpidsadnyi** [comment]
Unlike imperative context the dependency injection is statically declared. It makes a difference when you want to address a "customer" in the follow-up steps by name.
In pytest-bdd the given step can be resolved by its pythonic name if you want to like request.getfuncargvalue("customer") or simply requesting the customer fixture.
In imperative contexts you have to take care how to feed your context with the customer e.g ctx.customer = customer
And you also have to take care that given steps are called in the certain order, what if one given step is using the result of another. So the ordering, the naming matters...
I prefer dependency injection when each instance you are interested in has a name and by requesting it you create it implicitly.
The difference here is I don't think in ctx.add_whatever(role=X, name=Y)
I would say that customer is a model like class Customer()
and in this case @given('customer named (?p<name>\w+)'):
def customer_name(customer, name):
    customer.name = name

Or any other implementation of the customer__name first and then instantiating customer depending on its given name.

**Comment by olegpidsadnyi** [comment]
If customer in your case is a User with the role customer i'd say - rephrase to
And my user is a "customer" named "X"
if you need multiple instances please think of user, second_user, another_user and other conventions to identify the role of your entities in the specification. In any case you would have to identify "that guy" in the when or then section. It is not possible imho without a convention. Otherwise it is scripting. Pattern in the given section, pattern to address it in the when and then section. But imho specification of the behavior requires some static conventions not just values of the parameters.

**Comment by smoustaj** [comment]
Hi @olegpidsadnyi,

I've seen your comments and I'm not sure I understand your suggestions.
In my given statements I have to implement the following logics:
> Given I am on page Y
> And I insert X on the field Y
> And I insert Z on the field W

Or for example "I click on the button ADD" and "I click on the button ADD"
Obviously the "I insert X on the field Y" are pointing the same fixture. How would you solve it?

Thanks in advance!

**Comment by olegpidsadnyi** [comment]
@smoustaj I don't really understand your problem, but it seems you are using given section to take exercise. Given is 1 fact. You can't repeat it, it is given. For example "Given my article has 5 comments". If you repeat it 5 times do you expect 25 comments?
Exercise must be taken in the When section.

**Comment by smoustaj** [comment]
I'm posting you an example from the pytest-bdd pdf; where in the backgrounds section, they are treating this example: 

> Feature: Multiple site support
> Background:
> Given a global administrator named "Greg"
> And a blog named "Greg's anti-tax rants"
> And a customer named "Wilson"
> And a blog named "Expensive Therapy" owned by "Wilson"

In my case, i need to do this:

> Feature: Multiple site support
> 
> Background:
> Given a global administrator named "Greg"
> And a blog named "Greg's anti-tax rants"
> And a customer named "Wilson"
> And a blog named "Expensive Therapy" -owned by "Wilson"- 

As you note, i have used the BDD Step: "And a blog named {name} " twice; 
This doesn't work, is there another way to do it ??

Thank you

**Comment by olegpidsadnyi** [comment]
It seems you have 2 blogs. But in the Then section you would have to refer to that somehow. How do you make a difference between the first and the second?

In my projects I register fixtures literally like X and the second_X to be able to refer to that.

Then you are also implicitly mean that a blog belongs to Greg and the second one to Wilson.
You didn't specify the relation in your scenario/background.
If you make a convention of the second_ you could implement the relationship in python code, but it would be clear that they are related.

Given a customer named "Greg"
And a customer is a global administrator
And a blog named "Greg's anti-tax rants"
And a second customer named "Wilson"
And a second blog named "Expensive Therapy"

When I go to the second blog  # Assuming using credentials of the (first) customer? (which is also a global admin)
Then I should see "Hello World"

**Comment by sagism** [comment]
Reviving this issue, I'm bumping into the same problem, testing a tree-like data structure:

Given there is a root with value X
And there is a node with value Y
And there is a node with value Z
...

The second "And" step triggers that error.

Another case:
Given I have a garage named "Total Express"
And there is a car with license plate "123-678" and the make is Toyota
And its fuel tank has 32 gallons
And there is another car with license plate "456-333" and the make is Ford
And its fuel tank has 12 gallons
...

The "another" is superfluous, and I need to implement it separately, causing duplicate code.
The "its fuel tank has ... gallons" throws an error.
Yes, I could merge the "fuel tank" into the "there is a car" rule, but I prefer to keep those separate so I can reuse the "there is a car" clause in other scenarios where the fuel tank does not matter.

So bottom line, this limitation is annoying. At the very least add a configuration to relax it (perhaps even as part of the "strict gherkin" relaxing".
I don't mind fixing this... would it be okay?

**Comment by advance512** [comment]
Agreed, this is pretty annoying.

This is supported in `cucumber-js`, and I can attest to it making testing a lot easier and more fun.

**Comment by Nusnus** [comment]
I've also encountered it right now where the most intuitive option was to reuse the given statement but I got this error and I need to make a hack to make it work.
It could be nice if this will be supported in the near future :)

**Comment by feuillemorte** [comment]
Maybe we can add to given() function a flag (for example, reusable=False or smth like this) and add to documentation a big warning that in this case given() function ceases to be a fixture and become as a plain function?

**Comment by olegpidsadnyi** [comment]
How this idea would work with:
Given there's a book
And there's a book
When I read the book

What's book?

**Comment by olegpidsadnyi** [comment]
I'm still thinking imperative given is not a good idea.
When is the exercise for a given setup. Why setup should be an exercise? Why can't you say Given there are N books instead?

**Comment by feuillemorte** [comment]
It's up to developer how he will write it. But I'm thinking about this scenarios:
Given I'm starting a master node on 192.168.0.1
And I'm starting a worker node on 192.168.0.2

It might be in precondition of current test or a group of tests and I can't write it this way now

**Comment by olegpidsadnyi** [comment]
@feuillemorte a step represents some symbolic name of what is given so that you can refer to it later. Please mind the target_fixture feature of the step as well. In your scenario you have 2 steps for the master and for the worker, so there's no issue. Both can call the same function to start up the node.
In case you don't want to refer to the node by name later you can have a step
Given the nodes [192.168.0.1, 192.168.0.2] are running

Otherwise the setup looks like
Given I have 5 apples
And I have 2 apples
And I have 3 apples
When I eat 2 apples
There should be 8 apples left


This could be simply Given I have 10 apples

**Comment by olegpidsadnyi** [comment]
It is *pytest*-bdd, so the setup is done via fixtures

**Comment by feuillemorte** [comment]
> In your scenario you have 2 steps for the master and for the worker, so there's no issue. Both can call the same function to start up the node.

Yes, but it can looks like 
if node == 'master' ......
if node == 'worker' ......

so, you have 1 fixture for this and 2 values

if I use pytest I write with parametrized fixture or with some loop or smth else, doesn't matter. but if I use bdd it's more clear to see 
```
Given I'm starting a <master> node on <192.168.0.1>
And I'm starting a <worker> node on <192.168.0.2>
```
because in the end you will read this text.
Or you can create 1 function and 2-3-5-.. decorators given() which does one thing but named differently like
```
Given I'm starting a master node on 192.168.0.1
And I'm starting worker node on 192.168.0.3
And I'm starting another worker node on 192.168.0.3
And I'm starting the third worker node on 192.168.0.4
```
Maybe in this case much better to write like you
```
Given the nodes [192.168.0.1, 192.168.0.2] are running
```
but it depends on specific case and realization of test scenarios.
You can't advice everyone how he can write his current Given fixtures because everyone can think like me (that's why this task is opened) and he might think "if I can write it in when-and statement why I can't write it in given or background statement?". given() and when() looks the same, but given() is a fixture and when() is not.

So, maybe we can simply add this to documentation to clear situation and your point and close this task :)

**Comment by olegpidsadnyi** [comment]
I'm starting X sounds like When to me. Given something is running sounds like you want to visit it's URL or see something about it in the Then step. In order to do that you need a name to refer to. This is a perfect job for the fixture. In the When section you don't really care about the handler as a result of your action. So yeah. If you have a primitive that makes sure the node is running without any reference - I'd advice to make a universal one for multiple.
But I can easily imagine:
Given master is running on 192.168.0.1:8000
When I go to master node URL
And I stop the master URL
Then I should see 404

Maybe indeed this motivation has to be explained in the docs, but to me it is pretty much in the pytest spirit.

**Comment by kalfa** [comment]
I bumped in the same conceptual problem.

This is my conclusion (for the records so who's going to bump into it again might see it from another angle).

Pytest use fixtures, which are 'data'. The various verbs in the BDD feature file are mapped into fixutures when using pytest-bdd.

> Given Blog One
> And Blog Two

the expected result is:
- the result for the first line will be an object (i.e. data) representing Blog One
- he result for the fsecond line will be an object (i.e. data) representing Blog Two

and can be mapped into

> @given("Blog one"):
> def blog_one():
>    return BlogOne()
> 
> @given("Blog two"):
> def blog_two():
>    return BlogTwo()

and this works.

But, since there is the possibility of using parsers, there is the strong temptation of parametrising things and create

> @given(parse('Blog {name:s}')
> def create_block(name):
>     return Factory(name)


This turns the 'given' verb into a data-factory, not data. Which other BDD products manage (maybe?) this way. pytest-bdd cannot distinguish if we are talking about a data-factory fixture which returns data, or a data fixture itself.

One approach is to just consider fixtures as data.
The other approach is to allow parametrization and high flexibility creating those data-factories.
This is a matter of product's pholosophy and maybe BDD 'purity'.

The former avoids the misuse of the data associated to the verbs, since (AFAIK) the BDD feature syntax is a Product Owner (or whateve is called in the method you use) tool, not a generic testing tool.

The latter approach doesn't associate necessarily a BDD verb to a data, **but** the BDD verb creates the data. The data is stored for later retrieval somehow. There is association, but not strong coupling.
This can lead to a differnt use of BDD syntax, into really program-like steps which could be seen as a distorsion of the BDD ideas.

Aldo, from the technical point of view, the latter is incompatible the concept of fixtures in pytest and the fact that pytest-bdd couples fixtures with verb's data makes it not possible to have parametrised flexibility in it.

I fixture actually could return a function or a complex data structure. Could a fixture return fixtures?

anyway, all this to say: don't confuse data and data-factories. Both are cool, as long as we are aware those are different concepts

my 2p.

**Comment by MakersF** [comment]
I encountered this problem as well.

I see there are various opinions on the topic, and I agree with @olegpidsadnyi that the `given` should be data.

I see it as:
- `given`: the precondition (there is no action to do, this describes the state of the system)
- `when`: action under test
- `then`: expected post condition of the action

`given` being data is perfectly reasonable, and `pytest` gives a lot of flexibility on how to create that data.

The current behavior though makes it hard to describe a system which has a variable amount of data to create.

This is such a common need that `pytest` explicitly mentions it in it's own documentation: https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures

In this case a fixture is used as producer of objects, because tests might require a variable number of objects, or objects created with different parameters.

This is a very simple example

```
Given Alice is in chat A
  And Bob is in chat B
  And chat A auto-sync to chat B is enabled
When  Alice writes a message in chat A
Then  Bob sees a message in chat B
```

While trying to address the problem, I modified `pytest-bdd` to allow `given` to be used as a factory.
We can still access the data created by `given` with the fixture name in `then` and `when`, but we are getting a list which contains all the data created by the given, so that it can be referenced by the steps.

```python
@given(parsers.parse("{user} is in chat {chat}"), aggregate=True)
def user_in_chat(user, chat):
    return chat, user

@when(parsers.parse("{user} writes \"{message}\" in chat {chat}"))
def user_writes_in_chat(user, chat, message, chat_service, user_in_chat):
    # user_in_chat contains all the returned result of the given
    # In our example, both (Alice, A) and (Bob, B)
```

If you think this is a solution that fits the spirit of the project, I can look into cleaning up the PR and submit it. At the moment it's a fairly small code change, and it's backward compatible with the current behavior.

I think this is a very powerful feature, which brings `pytest-bdd` closer to feature parity with `behave`, plus all the power of `pytest`.

**Comment by kalfa** [comment]
> ```python
> @given(parsers.parse("{user} is in chat {chat}"), aggregate=True)
> def user_in_chat(user, chat):
>     return chat, user
> 
> @when(parsers.parse("{user} writes \"{message}\" in chat {chat}"))
> def user_writes_in_chat(user, chat, message, chat_service, user_in_chat):
>     # user_in_chat contains all the returned result of the given
>     # In our example, both (Alice, A) and (Bob, B)
> ```

thanks for looking into it.

What kind of object ```user_in_chat``` will be?

it seems in your implementation it's a kind of iterator which returns a pair (name, chat).
I wouldn't call it factory-data, but aggregated-data, but that's a detail :)
user_in_chat is about all the users and chats associations.

I think the next step is add lookup into this aggregated-data.

user_in_chat will be a opaque object which will be asked for data, according to the parameters the verb had.
I wouldn't really pass a list/tuple of 'bare' data, it exposes too much of implementation details and opens to testing problems.

In other words:
user_writes_in_chat shouldn't care how user_in_chat organised the data.
user_in_chat shouldn't care about organising it.
pytest-bdd is the man-in-the-middle which enable the communication by providing this opaque and querable object.

```python
@given(parsers.parse("{user} is in chat {chat}"), aggregate=True)
def user_in_chat(user, chat):
     return Chat(chat, user)
 
@when(parsers.parse("{user} writes \"{message}\" in chat {chat}"))
def user_writes_in_chat(user, chat, message, chat_service, user_in_chat):
    user_in_chat.get(user=user, chat=chat).write(message)
 ```

does it make sense?

I am not using pytest-bdd (I am answering becuase I like well organised code ;) anymore at the moment, but I'm curious about your implementation.

**Comment by MakersF** [comment]
As you say, the aggregation depends on the data and how the client wants to use it.

When `aggregate=True` it uses a list as a default option, as that makes it easy to use for simple cases.
But a custom type can be provided instead, which will be used for the aggregation, with `@given(parser, aggregate=MyAggregator)`

**Comment by anxuae** [comment]
Hello,

Could you consider to **relax** this constraint by using a parameter (like `strict_gherkin=False`)?

This is restriction is really painful. For instance: this step is to inject value in an automotive bus, different  named signals have to be sen in the same scenario:

```
@given(parsers.parse("BUS signal [{can_signal}] is equal to [{can_value:d}]"))
def ensure_bus_signal_value_is(can_signal, can_value, bus, context):
    ensure_bus_signal_value(can_signal, can_value, bus, context)
```
The BDD scenario is:

```
  Scenario: AC Button Value When Pressed
    Given BUS signal [NHVACACLampReq] is equal to [1]
    And BUS signal [NHVACACPowerReq] is equal to [2.6]
    When  I press air conditioning button for [3]s
    Then  air conditioning is starting
```

For me, the steps `    Given BUS signal [NHVACACLampReq] is equal to [1]` and `Given BUS signal [NHVACACPowerReq] is equal to [2.6]` are strictly different and should be authorized.

**Comment by Vic152** [comment]
I ran into this problem too. This is the design choice for the framework and it should be respected I guess. However, it does not mean it is not PITA. 

In my case I would like to reuse step from background in my scenario. And because I already used it, it throws the given step used exception.

There are ways around it. I could move tests to different feature file with its own background. Then some of my steps would land in fixtures as they need to be common now. I could have two implementations repeating code. I could use login_again=login and use login_again.

If framework ever allows for this, there is one point I would like to make. Let's not do it in `strict_gherkin=False`. I think this is a different issue and should be handled by different flag. I would like to enforce gherkin syntax but allow for reuse of the steps.

my 2p

**Comment by mdkess** [comment]
I ran into this as well. Running with some of the previous ideas, what if we had `aggregate=True` functions return a generator. Then we can still have predictable injection of other fixtures.

Taking @MakersF 's example from above:

```
Given Alice is in chat A
  And Bob is in chat B
  And chat A auto-sync to chat B is enabled
When Alice writes a message in chat A
Then Bob sees a message in chat B
```

You'd have something like this:

```python
# Some fixture that stores the data we need.
@pytest.fixture
def server():
  return Server()

# Because it has aggregate=True, returns a function that will be called
# for each given. Note the server fixture is injected.
@given('{username} is in chat {roomname}', aggregate=True)
def users_in_chatroom(server):
  def f(username, roomname):
    server.add_user_to_room_(username, roomname)
  return f

@given('chat {room1} auto-sync to chat {room2} is enabled')
def enable_auto_sync(server):
  server.enable_auto_sync(room1, room2)

@pytest.fixture
def message():
  return "some message"

@when('user {username} writes a message in room {roomname}')
def user_sends_message(username, roomname, message):
  server.send(username, roomname, message)

@then('user {username} sees message {message}')
def user_sees_message(server, username, message):
  assert message in server.get_messages_for_user(username)
```

**Comment by ThrashAbaddon** [comment]
This fixture related approach that @olegpidsadnyi  is talking about should be explained in the documentation. Total number of people stumbling over this just  indicates how much is it important that we all have this right from the start when we read the documentation.

**Comment by ThrashAbaddon** [comment]
@olegpidsadnyi So how would you solve using the fixtures the following problem? I'm just curious what is the pattern of solving this class of problems.

```py
import pytest
from pytest_bdd import scenario, given, when, then
from pytest_bdd.parsers import cfparse

import map_builder
from utils import create_vehicle, simulation_engine


@pytest.fixture
def map():
    return map_builder.Map()


@scenario('show.feature', 'Show 1')
def test_show_1(map):
    pass


@given(cfparse('there is {car:S} with speed {speed:f}', extra_types={'S': str}))
def _position_vehicle_on_a_road_and_set_speed(map, car, speed):
    car = create_vehicle(car)
    car.set_speed(speed)
    map.insert(car)


@when('we start the simulation')
def _run_the_simulation(map):
    simulation_engine(map)


@then('at least one of vehicles has collided')
def _check_for_collisions(map):
    assert map.collisions_exist()
```

```gherkin
Feature: Show

  Scenario: Show 1
    Given there is car 1 with speed 100
    And there is car 2 with speed 50
    And there is car 3 with speed -30
    And there is car 4 with speed 0
    When we start the simulation
    Then at least one of vehicles has collided
```

**Comment by olegpidsadnyi** [comment]
@ThrashAbaddon the fixture is a symbolic name that you refer to in the Then section. In this case this is "fire-and-forget" approach. This is the reason your then step is so vague. Like "at least one somewhere has collided, i'm not sure who with whom".

Most of the time testing is about one entity. But sometimes you have lists in your Ui. Or multiple objects in your case. What we do is we define multiple fixtures for that.
Most of the lists I can test with one, two, three items.
I define car (the most common case), second_car, third_car and other_car (in case it doesn't belong to this scene and has different parent).
Another thing I'm advocating against is declaring flavours. Not all the combinations can be described in one step as "Given there is a car with speed and color and doors and wheels and engine and .... and.. and..".

Instead we do:

Given the first car speed is 50
And the car is red
And the car position is 100, 100
And the car is heading North
And the second car speed is 100
And the second car position is 0, 100
And the second car is heading East
And there is a car
And there is a second car
When we start the simulation
Then the car is collided with the second car

These steps can be declared using target_fixture

@bdd.given(parse('the car speed is {speed}'), target_fixture='car__speed')
def given_car_speed(speed):
    return speed


I like the double underscore naming convention of FactoryBoy for the attributes.
This way you can declare your properties as fixtures and override them with the target_fixture or using @pytest.mark.parametrize("car__speed", [100])

You can also have a look at my other project https://github.com/pytest-dev/pytest-factoryboy
That allows you declaring factories for your models and register multiple instances (second, third etc.)
Then It will automatically explode them to fixtures with the model__attr conventional names.
So you don't have to think what you should override and simply guess.

**Comment by dcendents** [comment]
I also came here and after reading the whole thread, the solution is not obvious. But I finally get how to make it work by respecting the fact that in pytest-bdd, given is data and will become a pytest fixture. 

Maybe a new section could be added to the documentation with the following example (or your own).

Let's say I want to test a cli application, I could want to write the following scenario

```gherkin
Scenario: display the help message
    Given the cli parameter my_command
    And the cli parameter --help
    When I execute the cli
    Then I see the help message
```
With the code:

```python
@given(parsers.parse('the cli parameter {cli_param}'))
...
```
This will fail in pytest-bdd as the given method would be executed twice and is not allowed.

I can think of 2 ways to solve that.

---

First possibility, combine the 2 given statements together

```gherkin
Scenario: display the help message
    Given the cli parameters my_command, --help
    When I execute the cli
    Then I see the help message
```
with the code
```python
@given(parsers.cfparse('the cli parameter {cli_param:str+}', extra_types=dict(str=str)))
...
```
---

Second possibility, convert the given statements into when statements. Once I got past the feeling it is wrong to have multiple when statements (which purists will probably say), it made my tests much easier to write

```gherkin
Scenario: display the help message
    Given a cli parameter list
    When I add the cli parameter my_command
    And I add the cli parameter --help
    And I execute the cli
    Then I see the help message
```
```python
@given('a cli parameter list')
def cli_parameter_list():
...

@when(parsers.parse('I add the cli parameter {cli_param}'))
def add_cli_parameter(cli_parameter_list, cli_param):
...
```

I hope this can help other people getting stuck with this issue.

**Comment by DrewDevereux** [comment]
I too am hitting this issue. It is needed for when working with collections of objects; and in such cases there is no issue of ambiguity. I want to say:
```
Given I have three foos
And foo 1 is off
And foo 2 is off
And foo 3 is off
When I turn foo 2 on
Then foo 1 should be off
And foo 2 should be on
And foo 3 should be off
```
There is no problem of ambiguity here: the "I have {count:d} foos" step fixture can return a list of foos, and the "foo {n:d} is off" step can inspect or operate upon that list to ensure that the nth Foo is off.

Instead, I am forced to implement it as
```
Given I have Foos
    | 1 | off |
    | 2 | off |
    | 3 | off |
...
```
Which rather defeats the purpose of Gherkin being human-readable.

**Comment by nigelmegitt** [comment]
The fix implemented in #376 (and now unfortunately deployed in 4.0.0, which is where I first learned of it) surely breaks many more tests than it fixes, and seems like an absurd solution to the problem raised in this issue.

Semantically, it makes no sense at all to say that a Given step does not define a fixture. If there is one place where a step _should_ define a fixture, it is in a Given step!

I thought https://github.com/pytest-dev/pytest-bdd/issues/157#issuecomment-466396118 made a much more reasonable proposal that would not have been a breaking change and would have solved the problem.

Since the style of BDD that people here wanted to write did not work before, the release did not break any working tests for those guys in 4.0.0; however since there are many other working tests out there in other repos, that will have broken as a consequence of #376, and as @The-Compiler [noted](https://github.com/pytest-dev/pytest-bdd/issues/385#issuecomment-688508812), it is not clear that #376 actually resolves the issue anyway, I wonder if there is scope for rolling back #376 and reconsidering? Or making the given-step-not-a-fixture option one that can be specified by an option on the step definition instead of making it the default?

**Comment by olegpidsadnyi** [comment]
@nigelmegitt yes, 4.0 is a breaking change. That's why it is a major release number. The given step still can be a fixture, but the step function can't be used for that anymore.
As the readme suggests you can still define the fixture name explicitly.
```
@given("there's an article", target_fixture="article")
def there_is_an_article():
    return Article()
```

In case you want a dependency you can add it to the function signature (as the rest of the pytest does). Especially there could be many more dependencies.
```
@given("there's an article")
def there_is_an_article(article):
    pass
```

See
https://github.com/pytest-dev/pytest-bdd#migration-of-your-tests-from-versions-3xx

**Comment by nigelmegitt** [comment]
@olegpidsadnyi just because you _can_ make breaking changes in major versions does not mean you _should_! By changing the default here pytest-bdd could have retained backwards compatibility while allowing those who want the different behaviour to specify it.

**Comment by olegpidsadnyi** [comment]
@nigelmegitt can you describe your usage of the given step as a fixture? I have a VERY large test codebase in our project. It took me a day to convert the tests. In fact this discussion was out there for couple of years. I was even opposing the change. But when I looked the way they are used I couldn't really prove the given step function being an implicit fixture does any good. It may be decorated multiple times, it may clash with the fixture function implementing that setup. And we already had this target_fixture that does the explicit setup. Pytest-BDD was there for quite a while now. Sometimes it is also necessary to break the compatibility for the sake of the code quality and the design behind it.
I'm also using unit tests side-by-side with the functional tests. We are trying to use the declarative style as much as possible to get the most out of it. For example this project pytest-factoryboy helps a lot.
I understand the frustration of the breaking change. But besides that is there anything wrong about the syntax above?
Could you give me some examples of your usage of the given as a fixture?

**Comment by nigelmegitt** [comment]
@olegpidsadnyi sure happy to explain our usage.

I do not follow all the discussions about potential changes in all the upstream packages I use - it is simply not practical to do so; the first I knew about this was when 4.0.0 was released. I can see it has been under discussion for a while, now, but I had no idea beforehand.

Here's an example of our usage:

```gherkin
Feature: Value types from 3350

  # SPEC-CONFORMANCE: R93
  Scenario: Valid colour values
    Given an xml file <xml_file>
    When it has tts:color attribute with value <color>
    Then document is valid

    Examples:
    | xml_file             | color           |
    | 3350_value_types.xml | white           |
...
```

and

```python
@given('an xml file <xml_file>')
def template_file(xml_file):
    cur_dir = os.path.dirname(os.path.abspath(__file__))
    j2_env = Environment(loader=FileSystemLoader(os.path.join(cur_dir, 'templates')),
                         trim_blocks=True)
    return j2_env.get_template(xml_file)

@when('it has tts:color attribute with value <color>')
def when_color(color, template_dict):
    template_dict['color'] = color

@then('document is valid')
def valid_doc(template_file, template_dict):
    xml_file = template_file.render(template_dict)
    document = EBUTT3Document.create_from_xml(xml_file)
    assert isinstance(document, EBUTT3Document)
```

In this case, the Given step is creating the fixture `template_file` which is then used downstream. This is exactly the behaviour that a Given should be providing, I would say.

In terms of dealing with the kind of problem identified in this issue, we sometimes define a small number of differently labelled Given steps for "a first xml file" and "a second xml file" etc. 

To give a better idea of the scale of the problem for us, with 4.0.0 we have 710 failed tests and 228 passed, but with 3.x.x all of those failed tests pass.

**Comment by olegpidsadnyi** [comment]
@nigelmegitt there are quite some problems for the step being a fixture. In this case it can participate the dependency injection of other fixtures or the test parametrisation. Gherkin steps are imperative, and the fixture makes it not imperative anymore. We must make sure the step is evaluated exactly when it is executed by the scenario. I have a lot more examples of the negative side effects.
Still I believe that it is not too hard to walk through your given steps that do return and specify the target_fixture.

```python
@given('an xml file <xml_file>', target_fixture='template_file')
```

We have like 6700 tests and I was able to migrate them relatively easy. If you are reusing the fixtures well you should have not too many given steps. It should be approximately the number of your data model types and some of their key attributes. If you are striking the attribute of the model you are probably already using the target_fixture parameter.

Again, maybe pytest-factoryboy can be something interesting to you. Scientists say the declarative style saves up to 60% of the code size.

**Comment by DrewDevereux** [comment]
This fixes all my use cases and any use case I can imagine; and it looks like a very elegant and sensible approach. Thanks!

**Comment by nigelmegitt** [comment]
@olegpidsadnyi I've given this some more thought and discussed with colleagues, and I understand the problem a bit better now, and the solution to it. Making Given steps always fixtures does indeed cause problems, and although it's going to be additional work for me to fix up my tests, it is a "step" (pun intended) in the right direction, so I withdraw my request to roll back the changes. Thank you for your time and explanations.

**Comment by youtux** [comment]
I guess we can close this issue now. Fixed by https://github.com/pytest-dev/pytest-bdd/pull/376.
